### PR TITLE
Make no_offset_view inferrable

### PR DESCRIPTION
**Pull request by timholy**
This changes the implementation so that the axis *value*s
do not affect the type of the output.

Technically this is a breaking change, but only at the level
of the returned types, not at the level of returned values.

### CODE REVIEW

**Comment by timholy** [comment]
Test failure is just a coverage failure. I'd recommend waiting on this until #185 is dealt with, and any other "non-new-feature" PRs. Then this might go along with any other new features (#187? #179?) for a 1.6 release.

**Comment by jishnub** [comment]
I've wondered before if one objective of `no_offset_view` was to reduce the number of wrappers around the parent? Otherwise type-stability is a good idea

**Comment by timholy** [comment]
I'm sure it is, but anyone who complains can specialize it for their array type. Conversely, we can never work around the instability.

**Comment by timholy** [comment]
Before merging, I plan to add one fallback: for unknown array types with `Base.OneTo` axes, just return the array. That should address most of https://github.com/JuliaArrays/OffsetArrays.jl/pull/189#issuecomment-761853800.

**Review by johnnychen94 — APPROVED** [review]
A good change, but I hope we could pend this PR for a few weeks to see if there are other breaking changes; as far as I know, Augmentor.jl is vulnerable to this change due to its over-verbose dispatches and tests.

Said that, if this PR does fix some known performance issue, please do merge and release it.

**Comment by timholy** [comment]
> as far as I know, Augmentor.jl is vulnerable to this change due to its over-verbose dispatches and tests.

On 1.6, it seems to fail tests even on the master branch. Any further thoughts about this? There's no real urgency, but I wondered if the OneTo change here might circumvent Augmentor needing a fix.

**Comment by jishnub** [comment]
Perhaps `AbstractUnitRanges` may be transformed to other `AbstractUnitRange`s by `no_offset_view`? This should fix #198.

one choice could be 
```julia
OffsetArrays.no_offset_view(a::AbstractUnitRange) = UnitRange(a)
```

**Comment by timholy** [comment]
Better:

```julia
OffsetArrays.no_offset_view(a::AbstractUnitRange) = first(a):last(a)
```

That's more explicit that it only looks at the values, not the indices.

**Comment by jishnub** [comment]
Wouldn't this be open to someone defining a custom integer type such that `a:b` returns `Base.IdentityUnitRange(a:b)`? `UnitRange`s on the other hand always use `1`-based indexing.

**Comment by jishnub** [comment]
While we are at it, we might unwrap views of OffsetArrays as well. For example, 

```julia
julia> a = reshape(1:25, 5, 5);

julia> sa = @view a[2:3, 2:3];

julia> osa = OffsetArray(sa, OffsetArrays.Origin(2));

julia> sosa = @view osa[:,2];

julia> ososa = OffsetArrays.no_offset_view(sosa)
2-element OffsetArray(view(OffsetArray(view(reshape(::UnitRange{Int64}, 5, 5), 2:3, 2:3), 2:3, 2:3), :, 2), 1:2) with eltype Int64 with indices 1:2:
 7
 8

julia> av = @view sa[:, 1];

julia> av == ososa
true
```

In this case `no_offset_view(sosa)` could have stripped off a few layers and returned `av`. I wonder if such an optimization is worth it?

**Comment by timholy** [comment]
> Wouldn't this be open to someone defining a custom integer type such that a:b returns Base.IdentityUnitRange(UnitRange(a,b))? UnitRanges on the other hand always use 1-based indexing.


I guess I was worried about whether `UnitRange(Base.IdentityUnitRange(-1, 1))` should return `-1:1` or an error. `convert` seems like it should error (they are not "equivalent", whatever that means :smile:), but constructor-calls probably shouldn't.

**Comment by timholy** [comment]
> While we are at it, we might unwrap views of OffsetArrays as well

Good idea. I was a little bit worried about how far your suggestion delved into the internal representations of all types. What do you think about the more minimal version I just pushed? It doesn't doubly-unwrap like yours, but it does prevent the types from growing even more complex. As a bonus, it handles other types of shifted views, too (that's not piracy because we own `no_offset_view`). Demo:

```julia
julia> using OffsetArrays

julia> A = reshape(1:12, 3, 4)
3×4 reshape(::UnitRange{Int64}, 3, 4) with eltype Int64:
 1  4  7  10
 2  5  8  11
 3  6  9  12

julia> V = view(A, OffsetArrays.IdentityUnitRange(2:3), OffsetArrays.IdentityUnitRange(2:3))
2×2 view(reshape(::UnitRange{Int64}, 3, 4), :, :) with eltype Int64 with indices 2:3×2:3:
 5  8
 6  9

julia> typeof(V)
SubArray{Int64, 2, Base.ReshapedArray{Int64, 2, UnitRange{Int64}, Tuple{}}, Tuple{Base.IdentityUnitRange{UnitRange{Int64}}, Base.IdentityUnitRange{UnitRange{Int64}}}, false}

julia> OffsetArrays.no_offset_view(V)
2×2 view(reshape(::UnitRange{Int64}, 3, 4), 2:3, 2:3) with eltype Int64:
 5  8
 6  9

julia> typeof(OffsetArrays.no_offset_view(V))
SubArray{Int64, 2, Base.ReshapedArray{Int64, 2, UnitRange{Int64}, Tuple{}}, Tuple{UnitRange{Int64}, UnitRange{Int64}}, false}

julia> O = OffsetArray(A, -1:1, 0:3)
3×4 OffsetArray(reshape(::UnitRange{Int64}, 3, 4), -1:1, 0:3) with eltype Int64 with indices -1:1×0:3:
 1  4  7  10
 2  5  8  11
 3  6  9  12

julia> V = view(O, 0:1, 1:2)
2×2 view(OffsetArray(reshape(::UnitRange{Int64}, 3, 4), -1:1, 0:3), 0:1, 1:2) with eltype Int64:
 5  8
 6  9

julia> typeof(V)
SubArray{Int64, 2, OffsetMatrix{Int64, Base.ReshapedArray{Int64, 2, UnitRange{Int64}, Tuple{}}}, Tuple{UnitRange{Int64}, UnitRange{Int64}}, false}

julia> OffsetArrays.no_offset_view(V)
2×2 view(OffsetArray(reshape(::UnitRange{Int64}, 3, 4), -1:1, 0:3), 0:1, 1:2) with eltype Int64:
 5  8
 6  9

julia> typeof(OffsetArrays.no_offset_view(V))
SubArray{Int64, 2, OffsetMatrix{Int64, Base.ReshapedArray{Int64, 2, UnitRange{Int64}, Tuple{}}}, Tuple{UnitRange{Int64}, UnitRange{Int64}}, false}

julia> r1, r2 = OffsetArrays.IdOffsetRange(1:3, -2), OffsetArrays.IdentityUnitRange(2:3)
(OffsetArrays.IdOffsetRange(-1:1), Base.IdentityUnitRange(2:3))

julia> V = view(O, r1, r2)
3×2 view(OffsetArray(reshape(::UnitRange{Int64}, 3, 4), -1:1, 0:3), OffsetArrays.IdOffsetRange(-1:1), :) with eltype Int64 with indices OffsetArrays.IdOffsetRange(-1:1)×2:3:
 7  10
 8  11
 9  12

julia> typeof(V)
SubArray{Int64, 2, OffsetMatrix{Int64, Base.ReshapedArray{Int64, 2, UnitRange{Int64}, Tuple{}}}, Tuple{OffsetArrays.IdOffsetRange{Int64, UnitRange{Int64}}, Base.IdentityUnitRange{UnitRange{Int64}}}, false}

julia> OffsetArrays.no_offset_view(V)
3×2 view(OffsetArray(reshape(::UnitRange{Int64}, 3, 4), -1:1, 0:3), -1:1, 2:3) with eltype Int64:
 7  10
 8  11
 9  12

julia> typeof(OffsetArrays.no_offset_view(V))
SubArray{Int64, 2, OffsetMatrix{Int64, Base.ReshapedArray{Int64, 2, UnitRange{Int64}, Tuple{}}}, Tuple{UnitRange{Int64}, UnitRange{Int64}}, false}
```

To me this feels like a good place to be---I don't think we gain a lot by unwrapping the parent OffsetArray, and the risk is less generic handling of the axes.

**Comment by timholy** [comment]
So weird that codecov marks all of `no_offset_view` and its helpers as untested. Locally, every line of this diff is tested.

**Comment by jishnub** [comment]
> > While we are at it, we might unwrap views of OffsetArrays as well
> 
> Good idea. I was a little bit worried about how far your suggestion delved into the internal representations of all types. What do you think about the more minimal version I just pushed? It doesn't doubly-unwrap like yours, but it does prevent the types from growing even more complex. As a bonus, it handles other types of shifted views, too (that's not piracy because we own `no_offset_view`). Demo:
> 
> ```julia
> julia> using OffsetArrays
> 
> julia> A = reshape(1:12, 3, 4)
> 3×4 reshape(::UnitRange{Int64}, 3, 4) with eltype Int64:
>  1  4  7  10
>  2  5  8  11
>  3  6  9  12
> 
> julia> V = view(A, OffsetArrays.IdentityUnitRange(2:3), OffsetArrays.IdentityUnitRange(2:3))
> 2×2 view(reshape(::UnitRange{Int64}, 3, 4), :, :) with eltype Int64 with indices 2:3×2:3:
>  5  8
>  6  9
> 
> julia> typeof(V)
> SubArray{Int64, 2, Base.ReshapedArray{Int64, 2, UnitRange{Int64}, Tuple{}}, Tuple{Base.IdentityUnitRange{UnitRange{Int64}}, Base.IdentityUnitRange{UnitRange{Int64}}}, false}
> 
> julia> OffsetArrays.no_offset_view(V)
> 2×2 view(reshape(::UnitRange{Int64}, 3, 4), 2:3, 2:3) with eltype Int64:
>  5  8
>  6  9
> 
> julia> typeof(OffsetArrays.no_offset_view(V))
> SubArray{Int64, 2, Base.ReshapedArray{Int64, 2, UnitRange{Int64}, Tuple{}}, Tuple{UnitRange{Int64}, UnitRange{Int64}}, false}
> 
> julia> O = OffsetArray(A, -1:1, 0:3)
> 3×4 OffsetArray(reshape(::UnitRange{Int64}, 3, 4), -1:1, 0:3) with eltype Int64 with indices -1:1×0:3:
>  1  4  7  10
>  2  5  8  11
>  3  6  9  12
> 
> julia> V = view(O, 0:1, 1:2)
> 2×2 view(OffsetArray(reshape(::UnitRange{Int64}, 3, 4), -1:1, 0:3), 0:1, 1:2) with eltype Int64:
>  5  8
>  6  9
> 
> julia> typeof(V)
> SubArray{Int64, 2, OffsetMatrix{Int64, Base.ReshapedArray{Int64, 2, UnitRange{Int64}, Tuple{}}}, Tuple{UnitRange{Int64}, UnitRange{Int64}}, false}
> 
> julia> OffsetArrays.no_offset_view(V)
> 2×2 view(OffsetArray(reshape(::UnitRange{Int64}, 3, 4), -1:1, 0:3), 0:1, 1:2) with eltype Int64:
>  5  8
>  6  9
> 
> julia> typeof(OffsetArrays.no_offset_view(V))
> SubArray{Int64, 2, OffsetMatrix{Int64, Base.ReshapedArray{Int64, 2, UnitRange{Int64}, Tuple{}}}, Tuple{UnitRange{Int64}, UnitRange{Int64}}, false}
> 
> julia> r1, r2 = OffsetArrays.IdOffsetRange(1:3, -2), OffsetArrays.IdentityUnitRange(2:3)
> (OffsetArrays.IdOffsetRange(-1:1), Base.IdentityUnitRange(2:3))
> 
> julia> V = view(O, r1, r2)
> 3×2 view(OffsetArray(reshape(::UnitRange{Int64}, 3, 4), -1:1, 0:3), OffsetArrays.IdOffsetRange(-1:1), :) with eltype Int64 with indices OffsetArrays.IdOffsetRange(-1:1)×2:3:
>  7  10
>  8  11
>  9  12
> 
> julia> typeof(V)
> SubArray{Int64, 2, OffsetMatrix{Int64, Base.ReshapedArray{Int64, 2, UnitRange{Int64}, Tuple{}}}, Tuple{OffsetArrays.IdOffsetRange{Int64, UnitRange{Int64}}, Base.IdentityUnitRange{UnitRange{Int64}}}, false}
> 
> julia> OffsetArrays.no_offset_view(V)
> 3×2 view(OffsetArray(reshape(::UnitRange{Int64}, 3, 4), -1:1, 0:3), -1:1, 2:3) with eltype Int64:
>  7  10
>  8  11
>  9  12
> 
> julia> typeof(OffsetArrays.no_offset_view(V))
> SubArray{Int64, 2, OffsetMatrix{Int64, Base.ReshapedArray{Int64, 2, UnitRange{Int64}, Tuple{}}}, Tuple{UnitRange{Int64}, UnitRange{Int64}}, false}
> ```
> 
> To me this feels like a good place to be---I don't think we gain a lot by unwrapping the parent OffsetArray, and the risk is less generic handling of the axes.

This seems like a good (and a pretty cool) solution. If there are specific performance enhancements possible we might look into them later.

**Comment by jishnub** [comment]
We might require some special handling of `Base.Slice` to avoid changing the `IndexStyle`:

```julia
julia> a = reshape(1:25, 5, 5);

julia> sa = @view a[:, :];

julia> osa = OffsetArray(sa, OffsetArrays.Origin(2));

julia> sosa = @view osa[:, :]; IndexStyle(sosa)
IndexLinear()

julia> OffsetArrays.no_offset_view(sosa) |> IndexStyle
IndexCartesian()
```

File: src/OffsetArrays.jl [hunk 1]
```diff
@@ -436,7 +436,9 @@ julia> A
 """
 no_offset_view(A::OffsetArray) = no_offset_view(parent(A))
 no_offset_view(a::AbstractUnitRange) = UnitRange(a)
-no_offset_view(a::Base.Slice) = Base.Slice(UnitRange(a))
+if isdefined(Base, :IdentityUnitRange)
```
Comment (jishnub) [inline src/OffsetArrays.jl:439]: Why is this added only for julia versions > v1.0?

Comment (timholy) [reply src/OffsetArrays.jl:439]: Because `Slice` means 2 things on 1.0: "the whole axis" and "an IdentityUnitRange". The latter means that adding this for Julia 1.0 would be a bug.

Comment (jishnub) [reply src/OffsetArrays.jl:439]: In that case is there another way to avoid changing the indexstyle?

Comment (timholy) [reply src/OffsetArrays.jl:439]: No.

**Comment by timholy** [comment]
OK to merge?

File: src/OffsetArrays.jl (hunk 1 above)
Comment (jishnub) [reply src/OffsetArrays.jl:439]: In that case perhaps we may unwrap the `SubArray` only for versions >1.0? This way the `IndexStyle` is preserved on all versions.

Comment (timholy) [reply src/OffsetArrays.jl:439]: Done.

File: test/runtests.jl [hunk 2]
```diff
@@ -1415,21 +1416,59 @@ Base.axes(x::PointlessWrapper) = axes(parent(x))
...
+    end
+    O = OffsetArray(A, -1:1, 0:3)
+    V = view(O, 0:1, 1:2)
+    @test V == OffsetArrays.no_offset_view(V) == [5 8; 6 9]
+    r1, r2 = OffsetArrays.IdOffsetRange(1:3, -2), OffsetArrays.IdentityUnitRange(2:3)
+    V = view(O, r1, r2)
+    @test V != collect(V)
+    @test OffsetArrays.no_offset_view(V) == collect(V)
+    V = @view O[:,:]
+    if isdefined(Base, :IdentityUnitRange)
```
Comment (jishnub) [inline test/runtests.jl:1468]: This test may be carried out on v1.0 too, now that `SubArray`s are not unwrapped

Comment (timholy) [reply test/runtests.jl:1468]: Excellent catch!

**Comment by jishnub** [comment]
Looks good to be merged

**Comment by timholy** [comment]
Thanks for all your help, @jishnub!

**Comment by timholy** [comment]
Should this be 1.6 or 1.5.4? It's not exactly a bugfix, but neither is it a new feature.

**Comment by jishnub** [comment]
Perhaps 1.6, as certain return types are altered
